### PR TITLE
Add checkpoint field

### PR DESCRIPTION
**Pull request by nsinkov**
## Changes proposed in this PR

- add checkpoint field to job

## Why are we making these changes?

The checkpoint field will contain configuration to facilitate checkpointing in a job running on kubernetes with a custom shell that supports it. Useful when running on preemptable compute

### CODE REVIEW

File: scheduler/src/cook/schema.clj [hunk 1]
```diff
@@ -488,6 +494,34 @@ for a job. E.g. {:resources {:cpus 4 :mem 3} :constraints {\"unique_host_constra
...
+    :db/isComponent true
+    :db/cardinality :db.cardinality/one
+    :db.install/_attribute :db.part/db}
+   {:db/id (d/tempid :db.part/db)
+    :db/ident :checkpoint-options/preserve-paths
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/many
+    :db.install/_attribute :db.part/db}
+   {:db/id (d/tempid :db.part/db)
+    :db/ident :checkpoint-periodic-options/period-sec
```
Comment (dposada) [inline scheduler/src/cook/schema.clj:521]: Does this automatically make it a sub-attribute of `:checkpoint/periodic-options`?

File: scheduler/test/cook/test/rest/api.clj [hunk 2]
```diff
@@ -1392,6 +1393,26 @@
...
+          (let [conn (restore-fresh-database! "datomic:mem://mesos-api-test")
+                checkpoint {:mode "periodic"
+                            :options {:preserve-paths #{"p1" "p2"}}
+                            :periodic-options {:period-sec 777}}
+                {:keys [uuid] :as job} (assoc (minimal-job) :checkpoint checkpoint)]
+            (is (= {::api/results (str "submitted jobs " uuid)}
+                   (testutil/create-jobs! conn {::api/jobs [job]})))
+            (is (= (expected-job-map job)
+                   (dissoc (api/fetch-job-map (db conn) uuid) :submit_time)))))
+
```
Comment (dposada) [inline scheduler/test/cook/test/rest/api.clj:1415]: Should there be a test for `preemptible` mode?

File: integration/tests/cook/test_basic.py [hunk 3]
```diff
@@ -3006,6 +3006,26 @@ def test_kubernetes_disallowed_var_names(self):
...
+            command_enabled = 'bash -c \'if [[ "${COOK_CHECKPOINT_MODE}" == "auto" ]] && [[ "${COOK_CHECKPOINT_PERIOD_SEC}" == "555" ]] && [[ "${COOK_CHECKPOINT_PRESERVE_PATH_0}" == "p1" ]] && [[ "${COOK_CHECKPOINT_PRESERVE_PATH_1}" == "p2" ]]; then exit 0; else exit 1; fi\''
+            job_uuid_enabled, resp_enabled = util.submit_job(self.cook_url, command=command_enabled, container=container,
+                                                             checkpoint={"mode": "auto",
+                                                                         "periodic-options": {"period-sec": 555},
+                                                                         "options": {"preserve-paths": ["p2", "p1"]}})
+            self.assertEqual(201, resp_enabled.status_code)
+            util.wait_for_instance(self.cook_url, job_uuid_disabled, status='success')
+            util.wait_for_instance(self.cook_url, job_uuid_enabled, status='success')
+        finally:
+            util.kill_jobs(self.cook_url, [job_uuid_disabled, job_uuid_enabled])
```
Comment (dposada) [inline integration/tests/cook/test_basic.py:3027]: Should we add a test for the `periodic` and `preemptible` modes?

File: scheduler/docs/scheduler-rest-api.adoc [hunk 4]
```diff
@@ -83,6 +83,7 @@ The request body is list of job maps with each entry containing a map of the fol
 |`env` | JSON map | A map of environment variables to be provided to the job.
 |`constraints` | list | list of constraints in form [attribute, operator, pattern]
 |`disable_mea_culpa_retries` | boolean | Flag to disable mea culpa retries. If true, mea culpa retries will count against the job's retry count.
+|`checkpoint` | JSON map | optional configuration to enable checkpointing.
```
[the remark this answers is not in the document]
Comment (nsinkov) [reply scheduler/docs/scheduler-rest-api.adoc:86]: I can. it will be very generic...

File: scheduler/src/cook/schema.clj (hunk 1 above)
Comment (nsinkov) [reply scheduler/src/cook/schema.clj:521]: this does
`{:db/id (d/tempid :db.part/db)
    :db/ident :checkpoint/periodic-options
    :db/valueType :db.type/ref
    :db/isComponent true
    :db/cardinality :db.cardinality/one
    :db.install/_attribute :db.part/db}`
the  :db.type/ref :db/isComponent

File: scheduler/docs/scheduler-rest-api.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (nsinkov) [reply scheduler/docs/scheduler-rest-api.adoc:86]: I'd rather not try to cram it in here. adding the doc strings on the schema

File: scheduler/test/cook/test/rest/api.clj (hunk 2 above)
Comment (nsinkov) [reply scheduler/test/cook/test/rest/api.clj:1415]: I don't think so. No special options for it. Just a string

File: integration/tests/cook/test_basic.py (hunk 3 above)
Comment (nsinkov) [reply integration/tests/cook/test_basic.py:3027]: don't think so - just testing that the different types of elements of the configuration are working

**Pull request merged**